src/gdft.c

```diff
@@ -267,168 +260,172 @@ gdMutexDeclare (gdFontCacheMutex);
+				else
 					break;
-				}
-				*p++ = byte;
+
+				b0 = (uint8_t)str[i++];
+			}
+			if (b0 == suffix) {
+				*chPtr = n;
+				chPtr[1] = 0;
+				return i;
+			}
+		}
+	} else {
+		// string format alphanumeric
+		// copy str into buffer untill '\0' or ;
+		while (i<ENTITY_NAME_LENGTH_MAX) {
+			// not an entity
+			if (b0 == '\0')
+				return 0;
+			// have complete copy
+			if (b0 == suffix) {
+				entity_name_buf[n] = '\0';
+				break;
+			}
+			entity_name_buf[n++] = b0;
+			b0 = (uint8_t)str[i++];
+		}
+		// not an entity, too long
+		if (b0 != suffix)
+			return 0;
+		for (n = 0; n < NR_OF_ENTITIES; n++) {
+			if (strcmp((const char *)entities[n].name, (const char *)entity_name_buf) == 0) {
+				chPtr[0] = (uint32_t)entities[n].unicode_a;
+				chPtr[1] = (uint32_t)entities[n].unicode_b;
+				return i;
 			}
 		}
 	}
+	// no match
+	return 0;
+}
 
-	/*
-	 * Unroll 1 to 3 byte UTF-8 sequences, use loop to handle longer ones.
-	 */
-
-	byte = *((unsigned char *) str);
 #ifdef JISX0208
-	if (0xA1 <= byte && byte <= 0xFE) {
-		int ku, ten;
 
+static int gd_JISx0208_To_Unicode(const char *str, uint32_t *chPtr) {
+	int byte = *(uint8_t *)str;
+	int ku, ten;
+	if (0xA1 <= byte && byte <= 0xFE) {
 		ku = (byte & 0x7F) - 0x20;
 		ten = (str[1] & 0x7F) - 0x20;
 		if ((ku < 1 || ku > 92) || (ten < 1 || ten > 94)) {
-			*chPtr = (Tcl_UniChar) byte;
+			*chPtr = (uint32_t)byte;
 			return 1;
 		}
+	}
+	*chPtr = (uint32_t)UnicodeTbl[ku - 1][ten - 1];
+	return 2;
+}
 
-		*chPtr = (Tcl_UniChar) UnicodeTbl[ku - 1][ten - 1];
-		return 2;
-	} else
-#endif /* JISX0208 */
-		if (byte < 0xC0) {
-			/*
-			 * Handles properly formed UTF-8 characters between
-			 * 0x01 and 0x7F.  Also treats \0 and naked trail
-			 * bytes 0x80 to 0xBF as valid characters representing
-			 * themselves.
-			 */
-
-			*chPtr = (Tcl_UniChar) byte;
-			return 1;
-		} else if (byte < 0xE0) {
-			if ((str[1] & 0xC0) == 0x80) {
-				/*
-				 * Two-byte-character lead-byte followed
-				 * by a trail-byte.
-				 */
-
-				*chPtr = (Tcl_UniChar) (((byte & 0x1F) << 6) | (str[1] & 0x3F));
-				return 2;
-			}
-			/*
-			 * A two-byte-character lead-byte not followed by trail-byte
-			 * represents itself.
-			 */
-
-			*chPtr = (Tcl_UniChar) byte;
-			return 1;
-		} else if (byte < 0xF0) {
-			if (((str[1] & 0xC0) == 0x80) && ((str[2] & 0xC0) == 0x80)) {
-				/*
-				 * Three-byte-character lead byte followed by
-				 * two trail bytes.
-				 */
+#else
 
-				*chPtr = (Tcl_UniChar) (((byte & 0x0F) << 12)
-				                        | ((str[1] & 0x3F) << 6) | (str[2] & 0x3F));
-				return 3;
-			}
-			/*
-			 * A three-byte-character lead-byte not followed by
-			 * two trail-bytes represents itself.
-			 */
+static int gd_UTF8_To_Unicode(const char *str, uint32_t *chPtr) {
+	// reads the contents pointed to by str and outputs a unicode codepage into object pointed to by chPtr
+	// return is the bytes to offset for the first char of a subsequent read
```

Comment:
this is a bit hard to understand.  how about:

`// Returns the number of bytes read from |str|.`